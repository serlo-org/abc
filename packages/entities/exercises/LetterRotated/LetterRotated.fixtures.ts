import { Fixtures } from '../helpers';
import { IProps, IState } from './LetterRotated';

const props: IProps[] = [
  {
    letters: ['a', 'N', 'E'],
    rotated: [2],
    angles: ['50deg'],
    difficulty: 0.1
  },
  {
    letters: ['a', 'N', 'E', 's', 'T'],
    rotated: [2],
    angles: ['50deg'],
    difficulty: 0.3
  },
  {
    letters: ['a', 'N', 'E', 's', 'T'],
    rotated: [2, 3],
    angles: ['50deg', '-30deg'],
    difficulty: 0.5
  }
];

const fixtures: Fixtures<IProps, IState> = [
  {
    name: 'difficulty level 1 wrong',
    props: props[0],
    state: [true, false, false, false, false],
    isCorrect: false
  },
  {
    name: 'difficulty level 1 correct',
    props: props[0],
    state: [false, false, true, false, false],
    isCorrect: true
  },
  {
    name: 'difficulty level 2 wrong',
    props: props[1],
    state: [true, false, false, false, false],
    isCorrect: false
  },
  {
    name: 'difficulty level 2 correct',
    props: props[1],
    state: [false, false, true, false, false],
    isCorrect: true
  },
  {
    name: 'difficulty level 3 one wrong',
    props: props[2],
    state: [true, false, true, true, false],
    isCorrect: false
  },
  {
    name: 'difficulty level 3 one correct missing',
    props: props[2],
    state: [false, false, true, false, false],
    isCorrect: false
  },
  {
    name: 'difficulty level 3 correct',
    props: props[2],
    state: [false, false, true, true, false],
    isCorrect: true
  }
];

export default fixtures;
