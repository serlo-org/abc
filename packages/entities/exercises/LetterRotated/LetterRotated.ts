import { all, contains, map } from 'ramda';
import AbstractExercise from '../AbstractExercise';

export interface IProps {
  letters: string[];
  rotated: number[];
  angles: string[];
  difficulty: number;
}

export type IState = boolean[];

class LetterRotated extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return map(() => false, this.props.letters) as IState;
  }

  public isCorrect(selected: IState) {
    const { letters, rotated } = this.props;

    for (let i = 0; i < letters.length; i++) {
      if (selected[i] !== contains(i, rotated)) {
        return false;
      }
    }
    return true;
  }

  public isSubmitDisabled(selected: IState) {
    return all(sel => !sel, selected);
  }
}

export default LetterRotated;
