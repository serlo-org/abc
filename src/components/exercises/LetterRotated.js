import React, { Component } from 'react';
import { View } from 'react-native';

import { PRIMARY } from '../../styles/colors';
import { RoundTextButton } from '../Components';

class LettersRotated extends Component {
  constructor(props) {
    super(props);

    this.state = {
      highlighted: null
    };
  }

  createLetterButton = index => {
    const { angle, letters, rotated } = this.props;

    return (
      <RoundTextButton
        onPress={() => {
          this.setState({
            highlighted: index
          });
        }}
        highlighted={index === this.state.highlighted}
        text={letters[index]}
        size={60}
        style={[
          {
            marginLeft: 5,
            marginRight: 5
          },
          index === rotated ? { transform: [{ rotate: angle }] } : {}
        ]}
      />
    );
  };

  render() {
    return (
      <View
        style={{
          flex: 1,
          backgroundColor: PRIMARY,
          alignItems: 'center'
        }}
      >
        <View
          style={{
            flexDirection: 'row',
            marginTop: 50,
            alignItems: 'center',
            justifyContent: 'center'
          }}
        >
          {this.createLetterButton(0)}
          {this.createLetterButton(1)}
          {this.createLetterButton(2)}
        </View>
        <View
          style={{
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'center'
          }}
        >
          {this.createLetterButton(3)}
          {this.createLetterButton(4)}
        </View>
      </View>
    );
  }
}

export default LettersRotated;
