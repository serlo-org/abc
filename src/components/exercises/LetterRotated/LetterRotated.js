import React, { Component } from 'react';
import { View } from 'react-native';

import RoundTextButton from '../../common/RoundTextButton';
import { PortraitScreenOrientation } from '../../helpers/screen-orientation';

class LetterRotated extends Component {
  getTransform = index => {
    const angleIndex = this.props.rotated.indexOf(index);
    return angleIndex !== -1 ? [{ rotate: this.props.angles[angleIndex] }] : [];
  };

  createLetterButton = index => {
    const { letters, showFeedback, feedback } = this.props;
    const wrong =
      showFeedback &&
      (feedback.wrongChoices[index] || feedback.missingCorrectChoices[index]);
    const correct = showFeedback && feedback.correctChoices[index];
    //  const missingCorrect = showFeedback && feedback.missingCorrectChoices[index];
    return (
      <RoundTextButton
        onPress={() => {
          this.props.setState(state => {
            state[index] = !state[index];
            return state;
          });
        }}
        wrong={wrong}
        correct={correct}
        highlighted={this.props.state[index]}
        text={letters[index]}
        size={60}
        style={[
          {
            marginLeft: 5,
            marginRight: 5
          }
        ]}
        textStyle={{ transform: this.getTransform(index) }}
      />
    );
  };

  render() {
    return (
      <PortraitScreenOrientation>
        <View
          style={{
            flex: 1,
            alignItems: 'center'
          }}
        >
          <View
            style={{
              flexDirection: 'row',
              marginTop: 50,
              alignItems: 'center',
              justifyContent: 'center'
            }}
          >
            {this.createLetterButton(0)}
            {this.createLetterButton(1)}
            {this.createLetterButton(2)}
          </View>
          {this.props.difficulty >= 0.2 && (
            <View
              style={{
                flexDirection: 'row',
                alignItems: 'center',
                justifyContent: 'center'
              }}
            >
              {this.createLetterButton(3)}
              {this.createLetterButton(4)}
            </View>
          )}
        </View>
      </PortraitScreenOrientation>
    );
  }
}

export default LetterRotated;
