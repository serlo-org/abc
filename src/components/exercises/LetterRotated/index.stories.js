import { storiesOf } from '@storybook/react-native';

import * as LetterRotated from '.';
import { createStoriesFromFixtures } from '../helpers';

const story = storiesOf('exercises/LetterRotated', module);

createStoriesFromFixtures(story, LetterRotated);
