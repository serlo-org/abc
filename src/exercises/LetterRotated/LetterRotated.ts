import { map, contains } from 'ramda';
import AbstractExercise from '../AbstractExercise';

export interface IProps {
  letters: string[];
  rotated: number[];
  angles: string[];
  difficulty: number;
}

export type IState = boolean[];

class LetterRotated extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return map(() => false, this.getProps().letters) as IState;
  }

  public isCorrect(state: IState) {
    const { rotated } = this.getProps();
    for (let i = 0; i < state.length; i++) {
      if (state[i] !== contains(i, rotated)) {
        return false;
      }
    }
    return true;
  }
}

export default LetterRotated;
