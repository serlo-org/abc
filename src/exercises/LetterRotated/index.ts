import Exercise from './LetterRotated';
import fixtures from './LetterRotated.fixtures';

export { Exercise, fixtures };
