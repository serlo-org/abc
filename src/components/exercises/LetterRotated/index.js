import { LetterRotated } from '../../../exercises';
import Component from './LetterRotated';

export const Exercise = LetterRotated.Exercise;
export const fixtures = LetterRotated.fixtures;
export { Component };
