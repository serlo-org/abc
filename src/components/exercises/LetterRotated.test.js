import React from 'react';

import LetterRotated from './LetterRotated';

import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const tree = renderer.create(
    <LetterRotated
      letters={['a', 'N', 'E', 's', 'T']}
      rotated={4}
      angle="50deg"
    />
  );
  expect(tree).toMatchSnapshot();
});
