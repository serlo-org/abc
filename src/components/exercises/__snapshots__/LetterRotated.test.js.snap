// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00b4d5",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 50,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#73dbff",
            "borderRadius": 9999,
            "elevation": 5,
            "height": 60,
            "justifyContent": "center",
            "marginLeft": 5,
            "marginRight": 5,
            "opacity": 1,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
            "width": 60,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "norddruck",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          a
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#73dbff",
            "borderRadius": 9999,
            "elevation": 5,
            "height": 60,
            "justifyContent": "center",
            "marginLeft": 5,
            "marginRight": 5,
            "opacity": 1,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
            "width": 60,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "norddruck",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          N
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#73dbff",
            "borderRadius": 9999,
            "elevation": 5,
            "height": 60,
            "justifyContent": "center",
            "marginLeft": 5,
            "marginRight": 5,
            "opacity": 1,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
            "width": 60,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "norddruck",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          E
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#73dbff",
            "borderRadius": 9999,
            "elevation": 5,
            "height": 60,
            "justifyContent": "center",
            "marginLeft": 5,
            "marginRight": 5,
            "opacity": 1,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
            "width": 60,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "norddruck",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          s
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#73dbff",
            "borderRadius": 9999,
            "elevation": 5,
            "height": 60,
            "justifyContent": "center",
            "marginLeft": 5,
            "marginRight": 5,
            "opacity": 1,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
            "transform": Array [
              Object {
                "rotate": "50deg",
              },
            ],
            "width": 60,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "norddruck",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          T
        </Text>
      </View>
    </View>
  </View>
</View>
`;
